lib/bundler/settings.rb

```diff
@@ -130,21 +134,41 @@ def with
     end
 
     # @local_config["BUNDLE_PATH"] should be prioritized over ENV["BUNDLE_PATH"]
+    # Always returns an absolute path to the bundle directory
+    # TODO: Document and refactor this method
     def path
       key  = key_for(:path)
       path = ENV[key] || @global_config[key]
+      set_path = ""
+      install_path = ""
 
       if path && !@local_config.key?(key)
         path = "#{path}/#{Bundler.ruby_scope}" if path != Bundler.rubygems.gem_dir
-        return path
+        set_path = path
       end
 
       if path = self[:path]
         path = "#{path}/#{Bundler.ruby_scope}" if path != Bundler.rubygems.gem_dir
-        File.expand_path(path)
+        set_path = path
+      else
+        set_path = File.join(@root, Bundler.ruby_scope)
+      end
+
+      if set_path == File.join(Bundler.settings.root, Bundler.ruby_scope)
+        # ?
+        install_path = set_path
+      elsif set_path == Bundler.rubygems.gem_dir
+        # system gems path
+        install_path = set_path
+      elsif Pathname.new(set_path).absolute?
+        # all other absolute paths
+        install_path = set_path
       else
-        File.join(@root, Bundler.ruby_scope)
+         # all relative paths (configured by the user)
+        install_path = File.join(Bundler::root, Bundler.settings.path)
       end
+
+      install_path
```

Comment:
no need for this statement, as all the cases of the conditional end with the return value